Lombiq.Hosting.Tenants.Management/Controllers/ShellSettingsEditorController.cs

```diff
@@ -0,0 +1,85 @@
+﻿using Lombiq.Hosting.Tenants.Management.Constants;
+using Lombiq.Hosting.Tenants.Management.Models;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using Newtonsoft.Json;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Environment.Shell.Configuration;
+using OrchardCore.Modules;
+using OrchardCore.Mvc.Core.Utilities;
+using OrchardCore.Tenants.Controllers;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using static OrchardCore.Tenants.Permissions;
+
+namespace Lombiq.Hosting.Tenants.Management.Controllers;
+
+[Feature(FeatureNames.ShellSettingsEditor)]
+public class ShellSettingsEditorController : Controller
+{
+    private readonly IAuthorizationService _authorizationService;
+    private readonly IShellHost _shellHost;
+    private readonly SemaphoreSlim _tenantConfigSemaphore = new(1);
+    private readonly IShellConfigurationSources _tenantConfigSources;
+
+    public ShellSettingsEditorController(
+        IAuthorizationService authorizationService,
+        IShellHost shellHost,
+        IShellConfigurationSources tenantConfigSources)
+    {
+        _authorizationService = authorizationService;
+        _shellHost = shellHost;
+        _tenantConfigSources = tenantConfigSources;
+    }
+
+    [HttpPost]
+    [ValidateAntiForgeryToken]
+    public async Task<IActionResult> Edit(ShellSettingsEditorViewModel model)
+    {
+        if (!await _authorizationService.AuthorizeAsync(User, ManageTenants) ||
+            !_shellHost.TryGetSettings(model.TenantId, out var shellSettings))
+        {
+            return NotFound();
+        }
+
+        var settingsDictionary = JsonConvert.DeserializeObject<Dictionary<string, string>>(model.Json);
+        var newSettings = new Dictionary<string, string>();
+        foreach (var key in settingsDictionary.Keys)
+        {
+            if (shellSettings[key] != settingsDictionary[key])
+            {
+                var tenantSettingsPrefix = $"{model.TenantId}Prefix:{key}";
+                newSettings[tenantSettingsPrefix] = settingsDictionary[key];
+                newSettings[key] = settingsDictionary[key];
+            }
+        }
+
+        await _tenantConfigSemaphore.WaitAsync(HttpContext.RequestAborted);
```

Comment:
Locking for this is useful, but it'd be even better if it'd cover the multi-node hosting model too. See `IDistributedLock`.

Reply:
I'm using the same solution as it is in OC ShellSettingsManager.

Reply:
I'm really not sure if I understand this IDistributedLock, but from the docs in the code what I understood was that it should be used if multiple tenants can access something at the same time. If that's the case, this isn't a problem here. We are saving settings to the tenant config which is strictly one tenant related.

Reply:
It's not about multiple tenants, but multiple instances of a given tenant, like when the app is scaled out horizontally to two servers. The locking that's now here covers the issue of two people trying to save this at the same time on the same machine but won't help if this happens on two servers.

Reply:
Oh okay, then it is a possible scenario. I'll implement it.